taichi/ir/frontend_ir.cpp

```diff
@@ -721,28 +741,98 @@ bool IndexExpression::is_global() const {
   return is_field() || is_matrix_field() || is_ndarray();
 }
 
+static void field_validation(FieldExpression *field_expr,
+                             int index_dim,
+                             bool has_slice) {
+  TI_ASSERT(field_expr != nullptr);
+  if (field_expr->snode == nullptr)
+    return;
+  int field_dim = field_expr->snode->num_active_indices;
+
+  int has_dynamic = 0;
+  if (field_expr->snode->parent &&
+      field_expr->snode->parent->type == SNodeType::dynamic) {
+    has_dynamic = 1;
+  }
+  /*
+    Indexing to dynamic snode can be special, you can either index all the way
+    to place snode, or stop at the dynamic snode.
+  */
+  if (!has_slice && field_dim != index_dim &&
+      field_dim != has_dynamic + index_dim) {
+    throw TaichiTypeError(
+        fmt::format("Field with dim {} accessed with indices of dim {}",
+                    field_dim, index_dim));
+  }
+}
+
+static void matrix_field_validation(MatrixFieldExpression *matrix_field_expr,
+                                    int index_dim,
+                                    bool has_slice) {
+  TI_ASSERT(matrix_field_expr != nullptr);
+  TI_ASSERT(!matrix_field_expr->fields.empty());
+  auto field_expr = matrix_field_expr->fields[0].cast<FieldExpression>();
+  int element_dim = matrix_field_expr->element_shape.size();
+
+  if (field_expr->snode == nullptr)
+    return;
+  int outter_dim = field_expr->snode->num_active_indices;
+
+  int has_dynamic = 0;
+  if (field_expr->snode->parent &&
+      field_expr->snode->parent->type == SNodeType::dynamic) {
+    has_dynamic = 1;
+  }
+
+  /*
+    Indexing to dynamic snode can be special, you can either index all the way
+    to place snode, or stop at the dynamic snode.
+  */
+  if (!has_slice && outter_dim != index_dim &&
+      outter_dim != index_dim + has_dynamic &&
+      outter_dim + element_dim != index_dim &&
+      outter_dim + element_dim != index_dim + has_dynamic) {
+    throw TaichiTypeError(
+        fmt::format("Matrix with dim {} accessed with indices of dim {}",
+                    outter_dim + element_dim, index_dim));
+  }
+}
+
 void IndexExpression::type_check(CompileConfig *) {
   // TODO: Change to type-based solution
   // Currently, dimension compatibility check happens in Python
   TI_ASSERT(indices_group.size() == std::accumulate(begin(ret_shape),
                                                     end(ret_shape), 1,
                                                     std::multiplies<>()));
+  int index_dim = indices_group.empty() ? 0 : indices_group[0].size();
+  bool has_slice = !ret_shape.empty();
   if (!ret_shape.empty()) {
```

Comment:
The `has_slice` case has been checked here. So maybe we don't need the `has_slice` variable?